deployments/helm/gpud/templates/daemonset.yaml

```diff
@@ -96,31 +150,132 @@ spec:
+          # Privileged access is required for hardware inspection and low-level system access.
           securityContext:
             {{- toYaml .Values.securityContext | nindent 12 }}
 
           command:
-            {{- toYaml .Values.gpud.command | nindent 12 }}
-          args:
-            {{- toYaml .Values.gpud.args | nindent 12 }}
+            - /bin/sh
+            - -ec
+            - |
+              set -o errexit
+              set -o nounset
+              set -o pipefail 2>/dev/null || true  # pipefail not supported in dash (Ubuntu's /bin/sh)
+
+              # Node label file (if nodeLabelExporter sidecar is enabled)
+              LABEL_FILE="{{ .Values.nodeLabelExporter.outputFile }}"
+
+              # Label keys (can be overridden via env)
+              EP_LABEL_KEY="${GPUD_ENDPOINT_LABEL_KEY:-{{ .Values.nodeLabelExporter.labelKeys.endpoint }}}"
+              TOKEN_LABEL_KEY="${GPUD_TOKEN_LABEL_KEY:-{{ .Values.nodeLabelExporter.labelKeys.token }}}"
+              MID_LABEL_KEY="${GPUD_MACHINE_ID_LABEL_KEY:-{{ .Values.nodeLabelExporter.labelKeys.machineId }}}"
+              EP=""
+              TOKEN=""
+              MID=""
+              if [ -f "$LABEL_FILE" ]; then
+                # Use awk with index() for exact string matching to:
+                # 1. Avoid regex issues (dots/slashes in keys are literal)
+                # 2. Preserve '=' characters in values (important for base64 tokens)
+                EP="$(awk -v key="$EP_LABEL_KEY" 'index($0, key"=") == 1 {print substr($0, length(key)+2); exit}' "$LABEL_FILE")"
+                TOKEN="$(awk -v key="$TOKEN_LABEL_KEY" 'index($0, key"=") == 1 {print substr($0, length(key)+2); exit}' "$LABEL_FILE")"
+                MID="$(awk -v key="$MID_LABEL_KEY" 'index($0, key"=") == 1 {print substr($0, length(key)+2); exit}' "$LABEL_FILE")"
+              fi
+
+              # Build final command arguments using set -- for safety (avoids eval and quoting issues)
+              set -- run
+              set -- "$@" --listen-address="{{ .Values.gpud.listenAddress }}"
+              set -- "$@" --log-level="{{ .Values.gpud.logLevel }}"
+
+              # Determine endpoint (label takes precedence over values.yaml)
+              ENDPOINT="{{ .Values.gpud.endpoint }}"
+              if [ -n "$EP" ]; then
+                ENDPOINT="$EP"
+              fi
+              if [ -n "$ENDPOINT" ]; then
+                set -- "$@" --endpoint="$ENDPOINT"
+              fi
+              if [ -n "$TOKEN" ]; then
+                set -- "$@" --token="$TOKEN"
+              fi
+              if [ -n "$MID" ]; then
+                set -- "$@" --machine-id="$MID"
+              fi
+
+              set -- "$@" --enable-auto-update={{ .Values.gpud.enableAutoUpdate }}
+              set -- "$@" --auto-update-exit-code={{ .Values.gpud.autoUpdateExitCode }}
+
+              echo "Starting: /usr/local/bin/gpud $*"
+              exec /usr/local/bin/gpud "$@"
```

Comment:
This `exec` is so important to make signal handling/shutdown continue to work when switching to an inline shell script entrypoint.